modules/afprog/afprog.c

```diff
@@ -38,6 +38,33 @@
 #include <unistd.h>
 #include <signal.h>
 
+typedef struct _AFProgramReloadInfo
+{
+  LogWriter * writer;
+  pid_t       pid;
+} AFProgramReloadInfo;
+
+static inline void
+terminate_descendants_by_pid(const pid_t pid)
+{
+  msg_verbose("Sending descendant program group a TERM signal",
+              evt_tag_int("child_pid", pid),
+              NULL);
+
+  pid_t pgid = getpgid(pid);
+  if (pgid != -1)
+    killpg(pgid, SIGTERM);
+}
+
+static inline void
+afprogram_reload_info_free(AFProgramReloadInfo * const reload_info)
+{
+  child_manager_unregister(reload_info->pid);
+  terminate_descendants_by_pid(reload_info->pid);
+  log_pipe_unref(reload_info->writer);
+  g_free(reload_info);
+}
+
```

Comment:
please indicate somehow that this is not merely a free function, but rather a more complex beast that does something else than freeing. (killing processes is generally a lot more than a couple of frees)

maybe it could be split into two, a real _free function, and a method to tear down the rest of the state. The destroy_notify function could then be a private function related to the AFProgramDestDriver, which would call
the cleanup method and the free itself.